internal/exec/stages/files/filesystemEntries.go

```diff
@@ -16,16 +16,90 @@ package files
 
 import (
 	"fmt"
+	"io/ioutil"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"sort"
 	"strings"
 
 	"github.com/coreos/ignition/v2/config/v3_2_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/exec/util"
 	"github.com/coreos/ignition/v2/internal/log"
+
+	"github.com/vincent-petithory/dataurl"
 )
 
+// createCrypttabEntries creates entries inside of /etc/crypttab for LUKS volumes,
+// as well as copying keyfiles to the sysroot.
+func (s *stage) createCrypttabEntries(config types.Config) error {
+	if len(config.Storage.Luks) == 0 {
+		return nil
+	}
+
+	s.Logger.PushPrefix("createCrypttabEntries")
+	defer s.Logger.PopPrefix()
+
+	crypttab := fileEntry{
+		types.Node{
+			Path: "/sysroot/etc/crypttab",
+		},
+		types.FileEmbedded1{},
+	}
+	keyfiles := []fileEntry{}
+	for _, luks := range config.Storage.Luks {
+		out, err := exec.Command(distro.CryptsetupCmd(), "luksUUID", luks.Device).CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("gathering luks uuid: %s: %v", out, err)
+		}
+		uuid := strings.TrimSpace(string(out))
+		netdev := ""
+		if luks.Clevis != nil && len(luks.Clevis.Tang) > 0 {
+			netdev = ",_netdev"
+		}
+		keyfile := "none"
+		if luks.Clevis == nil {
+			keyfile = filepath.Join(util.LuksRealRootKeyFilePath, luks.Name)
+
+			// Copy keyfile from /run to sysroot
+			contents, err := ioutil.ReadFile(filepath.Join(util.LuksInitramfsKeyFilePath, luks.Name))
+			if err != nil {
+				return fmt.Errorf("reading keyfile for %s: %v", luks.Name, err)
+			}
+			contentsUri := dataurl.EncodeBytes(contents)
+			keyfiles = append(keyfiles, fileEntry{
+				types.Node{
+					Path: filepath.Join("/sysroot", keyfile),
```

Comment:
`s.JoinPath(keyfile)`.  Also mode 600.  The containing directory will be created mode 755; if we want mode 700, we'll need to do that explicitly.

Reply:
Are we concerned with the directory permissions?

Reply:
Restrictive permissions are often used for safety on config directories where nothing inside should be non-root-readable, and they'd also prevent users from seeing what keyfiles exist.  It'd be better to restrict the directory, but not required.

Reply:
That's fair, I'll punt for now and open up an issue once this lands.